kubeadm: cert renew doesn't work with `cert-dir` set

### What happened?

When `cert-dir` is set to a value other than the default value (/etc/kubernetes/pki/), and run `certs renew` on the working kubernetes controller node, the cert is not renewed as expected.

This impacts the `check-expiration` as well.




### What did you expect to happen?

`cert  renew` or `check-expiration` work as expected.

 

### How can we reproduce it (as minimally and precisely as possible)?

Here is an example which show cert renew doesn't work, 

- create a cert,
```bash
ls -l /tmp/3366554527.bak/apiserver.crt
-rw-r--r-- 1 root root 1277 Jun 20 18:33 /tmp/3366554527.bak/apiserver.crt
```

- and then renew the cert,

```
kubeadm certs renew apiserver --cert-dir /tmp/3366554527.bak

...
certificate for serving the Kubernetes API renewed
```

- check if the cert is renewed, 

```bash
ls -l /tmp/3366554527.bak/apiserver.crt
-rw-r--r-- 1 root root 1277 Jun 20 18:33 /tmp/3366554527.bak/apiserver.crt
```

It actually renewed the cert in the default path(/etc/kubernetes/pki/),
```bash
# ls -l /etc/kubernetes/pki/apiserver.crt
-rw-r--r-- 1 root root 1281 Jun 22 10:58 /etc/kubernetes/pki/apiserver.crt
```


The issue can be revealed by the test as well, 

```bash
go test k8s.io/kubernetes/cmd/kubeadm/app/cmd -test.run TestRunRenewCommands/apiserver
...
certificate the apiserver uses to access etcd renewed
--- FAIL: TestRunRenewCommands (0.13s)
    --- FAIL: TestRunRenewCommands/apiserver (0.03s)
        certs_test.go:256: certificate apiserver was not renewed as expected
    --- FAIL: TestRunRenewCommands/apiserver-kubelet-client (0.03s)
        certs_test.go:256: certificate apiserver-kubelet-client was not renewed as expected
    --- FAIL: TestRunRenewCommands/apiserver-etcd-client (0.02s)
        certs_test.go:256: certificate apiserver-etcd-client was not renewed as expected
FAIL
FAIL    k8s.io/kubernetes/cmd/kubeadm/app/cmd   0.158s
FAIL
```


**NOTE** This issue only exist with a working cluster since the code will fall back to use the cluster configuration when config file is not specified. 

And if the cluster cannot be connected with the `admin.conf`, the `cert-dir` will be honored, this is why the test can pass with the CI jobs.  Check the log printed by this test pr: https://github.com/kubernetes/kubernetes/pull/110681




### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
all
```

</details>


### Cloud provider

<details>
local
</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/cmd/kubeadm/app/cmd/certs.go
+++ b/cmd/kubeadm/app/cmd/certs.go
@@ -26,6 +26,7 @@ import (
 	"github.com/spf13/pflag"
 
 	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/klog/v2"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmscheme "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme"
@@ -335,6 +336,14 @@ func getInternalCfg(cfgPath string, kubeconfigPath string, cfg kubeadmapiv1.Clus
 			internalcfg, err := configutil.FetchInitConfigurationFromCluster(client, nil, logPrefix, false, false)
 			if err == nil {
 				fmt.Println() // add empty line to separate the FetchInitConfigurationFromCluster output from the command output
+				// certificate renewal or expiration checking doesn't depend on a running cluster, which means the CertificatesDir
+				// could be set to a value other than the default value or the value fetched from the cluster.
+				// cfg.CertificatesDir could be empty if the default value is set to empty (not true today).
+				if len(cfg.CertificatesDir) != 0 {
+					klog.V(1).Infof("Overriding the cluster certificate directory with the value from command line flag --%s: %s", options.CertificatesDir, cfg.CertificatesDir)
+					internalcfg.ClusterConfiguration.CertificatesDir = cfg.CertificatesDir
+				}
+
 				return internalcfg, nil
 			}
 			fmt.Printf("[%s] Error reading configuration from the Cluster. Falling back to default configuration\n\n", logPrefix)

